File: cpp/tests/test_visibility_astar.cpp
#include <cmath>
#include <limits>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/search/astar.hpp"
#include "navigation/global_planning/search/theta_star.hpp"
#include "navigation/global_planning/search/visibility_astar.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Cell;

namespace {

core::ParamSet cfg(const std::string& algo) {
  return core::ParamSet::from_yaml(test::repo_path("configs/global_planning/" + algo + ".yaml"));
}

double euclid(const Cell& a, const Cell& b) {
  double dr = a.row - b.row, dc = a.col - b.col;
  return std::sqrt(dr * dr + dc * dc);
}

// Every reconstructed any-angle leg must be a legal straight move.
bool path_los_clear(maps::OccupancyGrid2D& g, const std::vector<Cell>& path) {
  for (size_t i = 0; i + 1 < path.size(); ++i) {
    if (!g.line_of_sight(path[i], path[i + 1])) return false;
  }
  return true;
}

// Independent ground truth: the shortest path over the cell-centre visibility graph
// = Dijkstra over every pair of mutually LOS-visible reachable free cells, weighted
// by straight-line length. This is what Visibility A* optimises — a cell-centre
// approximation of the true any-angle optimum (turns pinned to cell centres, not
// obstacle corners). A different computation than the planner's interval search, so
// agreement is a real correctness check.
double visibility_optimum(maps::OccupancyGrid2D& grid, const Cell& start, const Cell& goal) {
  std::unordered_set<Cell> seen{start};
  std::vector<Cell> st{start};
  while (!st.empty()) {
    Cell c = st.back();
    st.pop_back();
    for (const auto& [nb, w] : grid.neighbors(c)) {
      (void)w;
      if (seen.insert(nb).second) st.push_back(nb);
    }
  }
  std::vector<Cell> cells(seen.begin(), seen.end());
  std::unordered_map<Cell, double> dist;
  dist[start] = 0.0;
  using QI = std::pair<double, Cell>;
  std::priority_queue<QI, std::vector<QI>, std::greater<QI>> pq;
  pq.push({0.0, start});
  std::unordered_set<Cell> settled;
  while (!pq.empty()) {
    auto [d, u] = pq.top();
    pq.pop();
    if (settled.count(u)) continue;
    settled.insert(u);
    if (u == goal) return d;
    for (const Cell& v : cells) {
      if (settled.count(v) || v == u) continue;
      if (grid.line_of_sight(u, v)) {
        double nd = d + euclid(u, v);
        auto it = dist.find(v);
        if (it == dist.end() || nd < it->second) {
          dist[v] = nd;
          pq.push({nd, v});
        }
      }
    }
  }
  return std::numeric_limits<double>::infinity();
}

}  // namespace

// (a) valid any-angle path ---------------------------------------------------

// Goal directly visible from start on an open grid: the cell-centre visibility
// optimum is the single straight segment (cost = Euclidean). Visibility A* returns
// exactly it, no longer than Theta* and strictly shorter than grid A*.
TEST(VisibilityAStar, ReturnsStraightLineWhenVisible) {
  auto g = test::make_grid({"...", "...", "..."});
  Cell start{2, 0}, goal{0, 1};
  global_planning::VisibilityAStarPlanner va(cfg("visibility_astar"));
  auto ra = va.plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_EQ(ra.path.front(), start);
  EXPECT_EQ(ra.path.back(), goal);
  EXPECT_NEAR(ra.cost, std::sqrt(5.0), 1e-9);
  EXPECT_TRUE(path_los_clear(g, ra.path));
  global_planning::ThetaStarPlanner theta(cfg("theta_star"));
  EXPECT_LE(ra.cost, theta.plan(g, start, goal, nullptr).cost + 1e-9);
  global_planning::AstarPlanner astar(cfg("astar"));
  EXPECT_LT(ra.cost, astar.plan(g, start, goal, nullptr).cost);
}

// A blocker hides the goal, forcing a bend. The cost must equal the cell-centre
// visibility-graph optimum and never exceed Theta*.
TEST(VisibilityAStar, MatchesVisibilityOptimumAroundObstacle) {
  auto g = test::make_grid({".....", ".....", "..#..", "..#..", "....."});
  Cell start{4, 0}, goal{0, 4};
  global_planning::VisibilityAStarPlanner va(cfg("visibility_astar"));
  auto ra = va.plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_EQ(ra.path.front(), start);
  EXPECT_EQ(ra.path.back(), goal);
  EXPECT_TRUE(path_los_clear(g, ra.path));
  EXPECT_FALSE(g.line_of_sight(start, goal));  // goal genuinely hidden
  EXPECT_NEAR(ra.cost, visibility_optimum(g, start, goal), 1e-9);
  global_planning::ThetaStarPlanner theta(cfg("theta_star"));
  EXPECT_LE(ra.cost, theta.plan(g, start, goal, nullptr).cost + 1e-9);
}

// Second, differently-shaped instance: guards against a projection correct only
// on the symmetric bend above.
TEST(VisibilityAStar, MatchesVisibilityOptimumOnAsymmetricMap) {
  auto g = test::make_grid({
      "........", "...##...", "...##...", ".....#..", "..#.....", "........",
  });
  Cell start{5, 0}, goal{0, 7};
  global_planning::VisibilityAStarPlanner va(cfg("visibility_astar"));
  auto ra = va.plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_TRUE(path_los_clear(g, ra.path));
  EXPECT_NEAR(ra.cost, visibility_optimum(g, start, goal), 1e-9);
  global_planning::ThetaStarPlanner theta(cfg("theta_star"));
  EXPECT_LE(ra.cost, theta.plan(g, start, goal, nullptr).cost + 1e-9);
}

// (b) no-path case -----------------------------------------------------------

TEST(VisibilityAStar, NoPathWhenWalledOff) {
  auto g = test::make_grid({"..#..", "..#..", "..#..", "..#..", "..#.."});
  global_planning::VisibilityAStarPlanner va(cfg("visibility_astar"));
  auto ra = va.plan(g, Cell{0, 0}, Cell{0, 4}, nullptr);
  EXPECT_FALSE(ra.success);
  EXPECT_TRUE(ra.path.empty());
  EXPECT_EQ(ra.cost, 0.0);
}

// (c) param validation failure -----------------------------------------------

TEST(VisibilityAStar, BadParamThrows) {
  std::string bad = test::write_temp(
      "visibility_astar.yaml",
      "algorithm: visibility_astar\ncategory: global_planning\nparams:\n"
      "  - name: heuristic_weight\n    type: float\n    default: 0.5\n    min: 1.0\n    max: 5.0\n"
      "    description: below min\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

File: python/tests/test_visibility_astar.py
"""Visibility A* — matches the cell-centre visibility-graph shortest cost, never
exceeds Theta*, and the no-path / parameter-validation contracts."""

from __future__ import annotations

import heapq
import io
import json
import math
from pathlib import Path

import pytest
from conftest import config, grid_from, open_grid, write_config

from navigation.core.params import ParamError, ParamSet
from navigation.core.trace import TraceRecorder
from navigation.core.types import Cell
from navigation.global_planning.search.astar import AStar
from navigation.global_planning.search.theta_star import ThetaStar
from navigation.global_planning.search.visibility_astar import VisibilityAStarPlanner


def _euclid(a: Cell, b: Cell) -> float:
    return math.hypot(a[0] - b[0], a[1] - b[1])


def _path_los_clear(grid: object, path: list[Cell]) -> bool:
    # Every reconstructed any-angle leg must be a legal straight move.
    for a, b in zip(path, path[1:], strict=False):
        if not grid.line_of_sight(a, b):  # type: ignore[attr-defined]
            return False
    return True


def _visibility_optimum(grid: object, start: Cell, goal: Cell) -> float:
    """Independent ground truth: the shortest path over the cell-centre visibility
    graph = Dijkstra over every pair of mutually LOS-visible reachable free cells,
    weighted by straight-line length. This is what Visibility A* optimises — a
    cell-centre approximation of the true any-angle optimum, since turns are pinned
    to cell centres rather than obstacle corners. Computed differently from the
    planner's interval search, so agreement is a real correctness check, not a
    restatement of the planner."""
    seen: set[Cell] = {start}
    stack = [start]
    while stack:  # reachable free component via 8-connected neighbors
        cur = stack.pop()
        for nb, _c in grid.neighbors(cur):  # type: ignore[attr-defined]
            if nb not in seen:
                seen.add(nb)
                stack.append(nb)
    cells = list(seen)
    dist: dict[Cell, float] = {start: 0.0}
    pq: list[tuple[float, Cell]] = [(0.0, start)]
    settled: set[Cell] = set()
    while pq:
        d, u = heapq.heappop(pq)
        if u in settled:
            continue
        settled.add(u)
        if u == goal:
            return d
        for v in cells:
            if v in settled or v == u:
                continue
            if grid.line_of_sight(u, v):  # type: ignore[attr-defined]
                nd = d + _euclid(u, v)
                if v not in dist or nd < dist[v]:
                    dist[v] = nd
                    heapq.heappush(pq, (nd, v))
    return math.inf


# (a) valid any-angle path ---------------------------------------------------

def test_visibility_astar_returns_straight_line_when_visible() -> None:
    # Goal directly visible from start on an open grid: the cell-centre visibility
    # optimum is the single straight segment, cost = Euclidean. Visibility A* must
    # return exactly it, no longer than Theta* and strictly shorter than grid A*.
    grid = open_grid(3, 3)
    start, goal = (2, 0), (0, 1)
    res = VisibilityAStarPlanner(config("visibility_astar")).plan(grid, start, goal)
    assert res.success
    assert res.path[0] == start and res.path[-1] == goal
    assert res.cost == pytest.approx(math.sqrt(5))  # analytically-known straight leg
    assert _path_los_clear(grid, res.path)
    theta = ThetaStar(config("theta_star")).plan(grid, start, goal)
    astar = AStar(config("astar")).plan(grid, start, goal)
    assert res.cost <= theta.cost + 1e-9
    assert res.cost < astar.cost


def test_visibility_astar_matches_visibility_optimum_around_obstacle() -> None:
    # A blocker hides the goal, forcing a bend. The cost must equal the cell-centre
    # visibility-graph optimum and never exceed Theta*, which only relaxes toward
    # its grandparent and can leave the string slightly slack.
    grid = grid_from([".....", ".....", "..#..", "..#..", "....."])
    start, goal = (4, 0), (0, 4)
    res = VisibilityAStarPlanner(config("visibility_astar")).plan(grid, start, goal)
    assert res.success
    assert res.path[0] == start and res.path[-1] == goal
    assert _path_los_clear(grid, res.path)
    assert not grid.line_of_sight(start, goal)  # goal genuinely hidden
    optimum = _visibility_optimum(grid, start, goal)
    assert res.cost == pytest.approx(optimum)
    theta = ThetaStar(config("theta_star")).plan(grid, start, goal)
    assert res.cost <= theta.cost + 1e-9


def test_visibility_astar_matches_visibility_optimum_on_asymmetric_map() -> None:
    # Second, differently-shaped instance: guards against a projection that is
    # correct only on the symmetric bend above.
    grid = grid_from([
        "........",
        "...##...",
        "...##...",
        ".....#..",
        "..#.....",
        "........",
    ])
    start, goal = (5, 0), (0, 7)
    res = VisibilityAStarPlanner(config("visibility_astar")).plan(grid, start, goal)
    assert res.success
    assert _path_los_clear(grid, res.path)
    assert res.cost == pytest.approx(_visibility_optimum(grid, start, goal))
    theta = ThetaStar(config("theta_star")).plan(grid, start, goal)
    assert res.cost <= theta.cost + 1e-9


# trace: (root, interval) info surfaced via the event data field ------------

def test_visibility_astar_emits_root_interval_data() -> None:
    buf = io.StringIO()
    VisibilityAStarPlanner(config("visibility_astar")).plan(
        open_grid(6, 6), (0, 0), (5, 5), TraceRecorder(buf)
    )
    intervals = [
        json.loads(line)["data"]
        for line in buf.getvalue().splitlines()
        if json.loads(line)["event"] == "edge_added"
    ]
    assert intervals, "must tag any-angle edges with their (root, interval) node"
    for d in intervals:
        assert d.keys() == {"row", "col_lo", "col_hi"}
        assert d["col_lo"] <= d["col_hi"]  # a real column run


# (b) no-path case -----------------------------------------------------------

def test_visibility_astar_no_path_when_walled_off() -> None:
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."])
    res = VisibilityAStarPlanner(config("visibility_astar")).plan(grid, (0, 0), (0, 4))
    assert not res.success
    assert res.path == []
    assert res.cost == 0.0


# (c) parameter validation failure -------------------------------------------

def test_visibility_astar_rejects_out_of_range_weight(tmp_path: Path) -> None:
    # visibility_astar declares heuristic_weight >= 1.0; a below-min default must
    # fail at load time, so an invalid weight can never reach plan().
    cfg = write_config(
        tmp_path / "bad_visibility_astar.yaml",
        "visibility_astar",
        [{"name": "heuristic_weight", "type": "float", "default": 0.5,
          "min": 1.0, "max": 5.0, "description": "below min"}],
    )
    with pytest.raises(ParamError):
        ParamSet.from_yaml(cfg)
